onnxruntime/contrib_ops/cpu/aten_ops/aten_op.cc

```diff
@@ -0,0 +1,76 @@
+// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License.
+
+#include "contrib_ops/cpu/aten_ops/aten_op.h"
+
+#include "core/dlpack/dlpack_converter.h"
+#include "core/framework/op_kernel_context_internal.h"
+#include "contrib_ops/cpu/aten_ops/aten_op_executor.h"
+
+namespace onnxruntime {
+namespace contrib {
+
+ONNX_OPERATOR_KERNEL_EX(ATen, kPytorchAtenDomain, 1, kCpuExecutionProvider,
+                        KernelDefBuilder().TypeConstraint("T", DataTypeImpl::AllTensorAndSequenceTensorTypes()), ATen);
+
+Status ATen::Compute(OpKernelContext* p_ctx) const {
+  auto* p_ctx_internal = static_cast<OpKernelContextInternal*>(p_ctx);
+  size_t input_size = static_cast<size_t>(p_ctx_internal->InputCount());
+  size_t output_size = static_cast<size_t>(p_ctx_internal->OutputCount());
+  std::unique_ptr<DLManagedTensor*[]> dlpack_inputs = std::make_unique<DLManagedTensor*[]>(input_size);
+  std::unique_ptr<DLManagedTensor*[]> dlpack_outputs = std::make_unique<DLManagedTensor*[]>(output_size);
+  for (size_t i = 0; i < input_size; ++i) {
+    const OrtValue* p_ort_value = p_ctx_internal->GetInputMLValue(static_cast<int>(i));
+    if (!p_ort_value) {
+      dlpack_inputs[i] = nullptr;
+    } else {
+      OrtValue ort_value = *p_ort_value;
+      dlpack_inputs[i] = dlpack::OrtValueToDlpack(ort_value);
+    }
+  }
+
+  aten_ops::ATenOperatorExecutor::Instance()(op_name_, overload_name_, input_size, dlpack_inputs.get(), output_size,
+                                             dlpack_outputs.get());
+  for (size_t i = 0; i < output_size; ++i) {
+    ORT_RETURN_IF_ERROR(
+        p_ctx_internal->SetOutputMLValue(static_cast<int>(i), dlpack::DlpackToOrtValue(dlpack_outputs[i])));
```

Comment:
This line of code looks very strange to me. How does it impact our memory planning/buffer reuse/etc?

Reply:
And if the output is the model's output, it totally doesn't work.

Reply:
The allocation planning is working just as the ATen Op is a regular Op. Normally an output tensor is allocated when the kernel calls OpKernelContext.Output. In ATen Op, we will not call OpKernelContext.Output, instead, We create the OrtValue directly (which has the ownership of the tensor from LibTorch) and set it the corresponding place in the ExecutionFrame. Anytime the ExecutionFrame decides to release the OrtValue, the tensor is also released. This is same for graph outputs. I changed the test models to expose ATen Op output as graph output, it also works. I am guessing the non-working case for graph output is those binding allocated graph outputs? I will try that. And I think the simple way is just not to support ATen op output as graph output, and fail such case in allocation planner.

Reply:
>  I changed the test models to expose ATen Op output as graph output, it also works

I think you didn't cover all the cases. If the output was allocated from outside(not managed by ORT), I think it will not work.

Reply:
In the latest commit I added the case to cover the external allocated outputs using binding output. Basically we have two ways:
- In SetOutputMLValue, if the OrtValue is allocated, perform a tensor copy (this is what I did in the commit)
- Or keep previous SetOutputMLValue, only set it when OrtValue is not allocated (assert if allocated), and modify the allocation plan to not allow ATen output as graph output by failing such case
Which one do you think is better?

Reply:
@souptc  what do you think?

Reply:
consider this "SetOutputMLValue" backdoor only exposed when build with ATenFallback, it probably ok we go with the first option here (make a copy if the output is already allocated, either by ort or external user). we can't afford to always copy as ATenFallback is heavily used in training, only copy when user using iobinding i think it is fine, as it is not very common when atenfallback and iobinding use together.